rfcs/0048-BYOC-Marvell-ML-accelerator-integration.md

```diff
@@ -0,0 +1,547 @@
+
+
+      # from Mrvl subgraph
+      %3 = @tvmgen_mrvl_main_3(%2, /* en_id=4139 */) /* ty=Tensor[(1, 14, 14, 32), float32] */;
+      def @tvmgen_mrvl_main_3(%mrvl_3_i0: Tensor[(1, 14, 14, 64), float32], Inline=1, Compiler="mrvl",
+          global_symbol="tvmgen_mrvl_main_3", Primitive=1) -> Tensor[(1, 14, 14, 32), float32] {
+
+        %13 = fn (%FunctionVar_0_0: Tensor[(1, 14, 14, 64), float32], PartitionedFromPattern="nn.conv2d_add_nn.relu_",
+            Composite="mrvl.conv2d_nhwc2nhwc") -> Tensor[(1, 14, 14, 32), float32] {
+          %11 = nn.conv2d(%FunctionVar_0_0, meta[relay.Constant][2] /* ty=Tensor[(32, 2, 2, 64), float32] */,
+              padding=[0, 0, 1, 1], channels=32, kernel_size=[2, 2], data_layout="NHWC", kernel_layout="OHWI",
+              out_layout="NHWC", /* en_id=781 */) /* ty=Tensor[(1, 14, 14, 32), float32] */;
+          %12 = add(%11, meta[relay.Constant][3] /* ty=Tensor[(1, 1, 1, 32), float32] */,
+              /* en_id=789 */) /* ty=Tensor[(1, 14, 14, 32), float32] */;
+          nn.relu(%12, /* en_id=793 */) /* ty=Tensor[(1, 14, 14, 32), float32] */
+        };
+
+        %13(%mrvl_3_i0, /* en_id=3343 */) /* ty=Tensor[(1, 14, 14, 32), float32] */
+      }
+```
+
+* Because Marvell backend layer uses NHWC format (for instance, for Conv2D, Pool2D, and Sum2D),
+    the relay.transform.ConvertLayout() pass is applied in the mrvl.py file. As a result, NHWC format is used
+    for Marvell layer: tvmgen_mrvl_main_1 to tvmgen_mrvl_main_4. In addition, the first tvmgen_mrvl_main_0 layer
+    is corresponding to a layout_transform() operation, which takes the original input tensor in src_layout="NCHW"
+    and convert the input to a dst_layout="NHWC" tensor.
+
+```
+      relay.transform.ConvertLayout(
+          {"nn.conv2d": ["NHWC", "OHWI"], "nn.max_pool2d": ["NHWC"]}
+      ),
+
+      %0 = @tvmgen_mrvl_main_0(%permute_input, /* en_id=4136 */) /* ty=Tensor[(1, 28, 28, 1), float32] */;
+      %1 = @tvmgen_mrvl_main_1(%0, /* en_id=4137 */) /* ty=Tensor[(1, 28, 28, 64), float32] */;
+      %2 = @tvmgen_mrvl_main_2(%1, /* en_id=4138 */) /* ty=Tensor[(1, 14, 14, 64), float32] */;
+      %3 = @tvmgen_mrvl_main_3(%2, /* en_id=4139 */) /* ty=Tensor[(1, 14, 14, 32), float32] */;
+      %4 = @tvmgen_mrvl_main_4(%3, /* en_id=4140 */) /* ty=Tensor[(1, 7, 7, 32), float32] */;
+
+      def @tvmgen_mrvl_main_0(%mrvl_0_i0: Tensor[(1, 1, 28, 28), float32], Inline=1, Compiler="mrvl",
+          global_symbol="tvmgen_mrvl_main_0", Primitive=1) -> Tensor[(1, 28, 28, 1), float32] {
+        layout_transform(%mrvl_0_i0, src_layout="NCHW", dst_layout="NHWC",
+            /* en_id=3334 */) /* ty=Tensor[(1, 28, 28, 1), float32] */
+      }
+```
+
+* Currently, in order for the following Marvell classes/functions to identify a Mrvl subgraphs and a non-Mrvl
+  subgraph from the layout-converted, composited/fused IR graph, we are utilizing the unique en_id attribute
```

Comment:
could you motivate the naming of `en_id` a bit? i recognize this is a common thing, but it might be nice to choose a slightly more specific name

Reply:
en_id as ExprNode ID. It is an extra field, which has been defined in the include/./tvm/ir/expr.h file for the RelayExprNode or just ExprNode class